### PR TITLE
support argmax converter

### PR DESCRIPTION
**Pull request by bowang007**
# Description

Support argmax converter

# Checklist:

- [ ] My code follows the style guidelines of this project (You can use the linters)
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests to verify my fix or my feature
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    dim: int = 0,
+    keep_dim: bool = False,
+) -> TRTTensor:
+    if not isinstance(input, TRTTensor):
+        raise RuntimeError(
+            f"argmax received input {input} that is not part " "of the TensorRT region!"
+        )
+    if input.dtype == trt.int32:
+        input = cast_trt_tensor(network, input, trt.float32, name)
+    if dim < 0:
```
[the remark this answers is not in the document]
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/impl/argmax.py:28]: Looks like we have to flatten the input first. 
Btw, I didn't find any converters related to `reshape` or `flatten`, maybe we should support that so we can use them directly.

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+from . import squeeze
+
+
+def argmax(
+    network: TRTNetwork,
+    target: Target,
+    source_ir: Optional[SourceIR],
+    name: str,
+    input: TRTTensor,
+    dim: int = 0,
```
[the remark this answers is not in the document]
Comment (bowang007) [reply py/torch_tensorrt/dynamo/conversion/impl/argmax.py]: I used `dim: Union[int, None]`, is that ok?

**Comment by bowang007** [comment]
Hey @gs-olive  I will be OOO next week. 
I think this update covers all edge cases. 
Please feel free to merge if this is good to go. Thanks!

**Review by gs-olive — APPROVED** [review]
Looks good to me

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    # 1. dim == None, flatten input tensor first, keep_dim will be ignore and the output rank == input rank
+    # 2. input rank == 1: TopK layer does not support 1 dimensional topk operation. Broadcast input to rank == 2
+    # 3. normal cases, no additional handlings
+    out = input
+
+    if dim is None:
+        shuffle_layer = ctx.net.add_shuffle(input)
+        shuffle_layer.reshape_dims = (*flatten_dims(input, 0, -1), 1)
+        set_layer_name(shuffle_layer, target, name + "_flatten")
+        out = shuffle_layer.get_output(0)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/argmax.py:43]: Consider using `reshape`:
```
new_shape = (*flatten_dims(input, 0, -1), 1)
out = impl.shuffle.reshape(ctx, target, source_ir, f"{name}_flatten", input, new_shape)
```

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    elif len(input.shape) == 1:
+        shuffle_layer = ctx.net.add_shuffle(input)
+        shuffle_layer.reshape_dims = (*input.shape, 1)
+        set_layer_name(shuffle_layer, target, name + "_broadcast")
+        out = shuffle_layer.get_output(0)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/argmax.py:48]: Same as above.

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+from torch.fx.node import Target
+from torch_tensorrt.dynamo._SourceIR import SourceIR
+from torch_tensorrt.dynamo.conversion._ConversionContext import ConversionContext
+from torch_tensorrt.dynamo.conversion.converter_utils import (
+    cast_trt_tensor,
+    flatten_dims,
+    get_axes_for_reduce_op,
+)
+from torch_tensorrt.fx.converters.converter_utils import (
+    get_positive_dim,
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/argmax.py:13]: Consider switching to `from torch_tensorrt.dynamo.conversion.converter_utils import get_positive_dim`

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        shuffle_layer.reshape_dims = (*input.shape, 1)
+        set_layer_name(shuffle_layer, target, name + "_broadcast")
+        out = shuffle_layer.get_output(0)
+
+    reduce_mask = get_axes_for_reduce_op(0)
+    if dim is not None:
+        reduce_mask = get_axes_for_reduce_op(get_positive_dim(dim, len(out.shape)))
+
+    topk_layer = ctx.net.add_topk(out, trt.TopKOperation.MAX, 1, reduce_mask)
+    set_layer_name(topk_layer, target, name)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/argmax.py:55]: Consider switching to `set_layer_name(topk_layer, target, name, source_ir)`

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    out = topk_layer.get_output(1)
+
+    if dim is None:
+        out_shuffle_layer = ctx.net.add_shuffle(out)
+        out_shuffle_layer.reshape_dims = (1,) * len(input.shape) if keep_dim else ()
+        set_layer_name(out_shuffle_layer, target, name + "_broadcast")
+        out = out_shuffle_layer.get_output(0)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/argmax.py:63]: Consider using `impl.shuffle.reshape`.

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    elif len(input.shape) == 1:
+        out = squeeze.squeeze(
+            ctx,
+            target,
+            SourceIR.ATEN,
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/argmax.py:68]: Consider `SourceIR.ATEN` -> `source_ir`

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            name + "_squeeze",
+            out,
+            1 if keep_dim else [0, 1],
+        )
+    elif not keep_dim:
+        out = squeeze.squeeze(ctx, target, SourceIR.ATEN, name + "_squeeze", out, dim)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/argmax.py:74]: Consider `SourceIR.ATEN` -> `source_ir`

**Review by zewenli98 — CHANGES_REQUESTED** [review]
Looks good overall, but I left some comments about using our new APIs and small fixes.

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+    ctx: ConversionContext,
+    target: Target,
+    source_ir: Optional[SourceIR],
+    name: str,
+    input: TRTTensor,
+    dim: Optional[int],
+    keep_dim: bool = False,
+) -> TRTTensor:
+    if input.dtype == trt.int32:
+        input = cast_trt_tensor(ctx, input, trt.float32, name, source_ir)
```
Comment (zewenli98) [inline py/torch_tensorrt/dynamo/conversion/impl/argmax.py:28]: It looks like missing `target`. The interface is:
```
def cast_trt_tensor(
    ctx: ConversionContext,
    input_val: TRTTensor,
    dtype: TRTDataType,
    name: str,
    target: Target = "",
    source_ir: Optional[SourceIR] = None,
) -> TRTTensor:
```

**Review by zewenli98 — COMMENTED** [review]
Thanks for the fix! I found a small bug here. Other looks good to me!

File: py/torch_tensorrt/dynamo/conversion/impl/argmax.py (hunk 10 above)
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/argmax.py]: Great catch - thanks!

**Review by zewenli98 — APPROVED** [review]
LGTM!

**Pull request merged**